src/platform/Darwin/DnssdImpl.cpp

```diff
@@ -215,20 +275,55 @@ static void OnGetAddrInfo(DNSServiceRef sdRef, DNSServiceFlags flags, uint32_t i
 {
     ChipLogProgress(Discovery, "Mdns: %s flags: %d, interface: %u, hostname: %s", __func__, flags, (unsigned) interfaceId,
                     StringOrNullMarker(hostname));
-    auto sdCtx = reinterpret_cast<ResolveContext *>(context);
+
+    auto contextWithType = reinterpret_cast<ResolveContextWithType *>(context);
+    VerifyOrReturn(contextWithType != nullptr, ChipLogError(Discovery, "ResolveContextWithType is null"));
+
+    auto sdCtx = reinterpret_cast<ResolveContext *>(contextWithType->context);
     ReturnOnFailure(MdnsContexts::GetInstance().Has(sdCtx));
     LogOnFailure(__func__, err);
 
     if (kDNSServiceErr_NoError == err)
     {
-        sdCtx->OnNewAddress(interfaceId, address);
+        InterfaceKey interfaceKey = { interfaceId, hostname, contextWithType->isSRPType };
+        sdCtx->OnNewAddress(interfaceKey, address);
+
+        // Set the flag to start the timer for resolve on SRP domain to complete if the key has the SRP type requested flag set to
+        // true.
+        if (interfaceKey.isSRPTypeRequested)
+        {
+            sdCtx->shoulStartSRPTimerForResolve = true;
+        }
+    }
+
+    if (flags & kDNSServiceFlagsMoreComing)
+    {
+        return;
     }
 
-    if (!(flags & kDNSServiceFlagsMoreComing))
+    VerifyOrReturn(sdCtx->HasAddress(), sdCtx->Finalize(kDNSServiceErr_BadState));
+
+    // We should complete the resolve if we got a resolution on non SRP domain and no resolution on SRP domain was requested.
+    if (!sdCtx->shoulStartSRPTimerForResolve)
     {
-        VerifyOrReturn(sdCtx->HasAddress(), sdCtx->Finalize(kDNSServiceErr_BadState));
         sdCtx->Finalize();
     }
+    else
+    {
+        // Since we started a resolve on the SRP domain we are going to start a timer to give the resolve some extra time to
```

Comment:
Where are we checking for "we got an SRP answer already, no need to start the timer"?

All this timer-management logic seems wrong.

Reply:
tried to fix it